### PR TITLE
wayland.windowManager.niri: add opt-in upstream defaults

### LINKED ISSUE #9741: [Feature Request] niri: Provide default values for `settings`
### Module Name

wayland.windowManager.niri

### Requested Feature

I would like to request either defining [Niri's default config](https://github.com/niri-wm/niri/blob/main/resources/default-config.kdl) as a Nix Attribute Set inside `wayland.windowManager.niri.settings`, or providing an alternative workaround to achieve a similar result.

I have considered two potential approaches. I would love to gather feedback, and if there are no major objections, I am willing to implement this and submit a PR.

#### 1. Define the upstream defaults directly within the options of `settings`

- **Pros**: Users can tweak specific parts of the config without losing the rest of the defaults.
- **Cons**: Maintenance overhead to keep the default values along with upstream specification changes. There is room for discussion on whether Home Manager strictly needs to keep them up-to-date without delay.

#### 2. Fallback to the upstream config file only when `settings` is empty

- **Pros**: Relatively easy to maintain.
- **Cons**: Not ideal for UX. As soon as a user defines even a single custom option, all other unmentioned defaults will disappear.

### Current Limitations

If `settings` is not defined and only `extraConfig` is populated, the generated config.kdl will contain only the contents written in `extraConfig`.
Comparing this to the conventions of existing modules such as [`wayland.windowManager.sway`](https://github.com/nix-community/home-manager/blob/e705714e918c3b11affcdd15db2cbe3a070420a0/modules/services/window-managers/i3-sway/sway.nix#L678-L682), this behavior might differ from what users would normally expect.

### Additional Context

CC: @lukasngl

### PR DESCRIPTION
**Pull request by glmlm**
### Description

Closes #9741

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `nix fmt` or
      `nix-shell -A dev --run treefmt`.

- [x] Code tested through `nix build .#test-all`
      or a targeted `nix run .#tests -- <pattern>`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

  ```
  {component}: {description}

  {long description}
  ```

  See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/a51598236f23c89e59ee77eb8e0614358b0e896c/modules/programs/lesspipe.nix#L11).
  - [ ] Generate a news entry. See [News](https://nix-community.github.io/home-manager/index.xhtml#sec-news)
  - [ ] Basic tests added. See [Tests](https://nix-community.github.io/home-manager/index.xhtml#sec-tests)

- If this PR adds an exciting new feature or contains a breaking change.
  - [ ] Generate a news entry. See [News](https://nix-community.github.io/home-manager/index.xhtml#sec-news)

### CODE REVIEW

File: modules/services/window-managers/niri.nix [hunk 1]
```diff
@@ -188,7 +188,412 @@ in
...
+            };
+            "Ctrl+Alt+Delete" = {
+              quit = { };
+            };
+            "Mod+Shift+P" = {
+              power-off-monitors = { };
+            };
+          };
+        };
+        settings = lib.trim (toKDL (lib.recursiveUpdate defaultSettings cfg.settings));
```
Comment (ambroisie) [inline modules/services/window-managers/niri.nix]: I'd tend to implement this in a similar manner to the i3/Sway module (using a default configuration), so that the user can easily:

1. Use `mkOptionDefault` to keep the default settings and selectively add/override some bindings.
2. _Not_ use `mkOptionDefault` and override the default bindings.

Comment (ambroisie) [reply modules/services/window-managers/niri.nix]: See https://github.com/nix-community/home-manager/blob/2e6715977106c498355a70bcefd2c78a33c94558/modules/services/window-managers/i3-sway/i3.nix#L44-#L123

Comment (lukasngl) [reply modules/services/window-managers/niri.nix]: I would +1 this, i don't use most of the default binds, so having the option to just override them wholesale or piecemeal seems like a good middle ground.

File: tests/modules/services/niri/niri-empty-config.nix [hunk 2]
```diff
@@ -10,7 +10,7 @@
     systemd.enable = false;
     xwaylandSatellitePackage = null;
 
-    # Empty Config should not generate `$XDG_CONFIG_HOME/niri/config.kdl`
+    # Providing empty config is expected to populate default values at `$XDG_CONFIG_HOME/niri/config.kdl`
```
Comment (lukasngl) [inline tests/modules/services/niri/niri-empty-config.nix]: Would you mind keeping a test case to assert that it's possible and trivial to produce an empty config? I personally would prefer to keep `builtins.readFile` my config from a kdl file instead of writing kdl flavoured nix.

**Review by lukasngl — CHANGES_REQUESTED** [review]
Hi, thanks for the contribution! I think it looks solid, please just move the default settings to the option's default (make sure to set defaultText aswell)

File: tests/modules/services/niri/niri-empty-config.nix (hunk 2 above)
Comment (glmlm) [reply tests/modules/services/niri/niri-empty-config.nix]: Thank you for your review. Updated the test case and reverted the expected output to the original.

File: modules/services/window-managers/niri.nix [hunk 3]
```diff
@@ -176,6 +172,414 @@ in
...
+              spawn = [ "swaylock" ];
+            };
+            "Super+Alt+S" = {
+              _props.allow-when-locked = true;
+              _props.hotkey-overlay-title = null;
+              spawn-sh = [ "pkill orca || exec orca" ];
+            };
+            "XF86AudioRaiseVolume" = {
+              _props.allow-when-locked = true;
+              spawn-sh = "wpctl set-volume @DEFAULT_AUDIO_SINK@ 0.1+ -l 1.0";
```
Comment (khaneliman) [inline modules/services/window-managers/niri.nix]: Some of these actions only exist in recent niri (`spawn-sh`, `toggle-overview`, `maximize-window-to-edges`), and `checkConfig` validates against `cfg.package`, so users on an older niri get a hard build failure from binds they never wrote. A baked snapshot is only correct for the exact niri version it was copied from.

File: modules/services/window-managers/niri.nix (hunk 1 above)
Comment (glmlm) [reply modules/services/window-managers/niri.nix]: I wasn't entirely sure of the best practice, so this might be a bit clunky... Anyway, it now correctly allows overriding settings with things like `lib.mkForce`.

**Edit**: I might have misunderstood it slightly. I didn't realize that `mkOptionDefault` is supposed to be used on the user side. In any case, I would appreciate it if you could review the current design. If there are any problems, I will move it to `default`.

Comment (khaneliman) [reply modules/services/window-managers/niri.nix]: I don't think the i3/sway `mkOptionDefault` pattern can work with this option's shape, unfortunately: the action is the attr *name*, so overriding `binds."Mod+Q".spawn = [ ... ]` merges with the default `close-window = { }` and produces a bind with two actions. You'd need `mkForce` per bind (or over all of `settings`) for any real customization.

niri >= 25.11 gives us a nicer way to get that same UX: [`include`](https://github.com/niri-wm/niri/wiki/Configuration%3A-Include) merges with the main config, and later `binds` override included keys individually. So instead of maintaining a snapshot here (which will drift from upstream the way our sway defaults have), we could extract the binds block from the user's own niri at build time:

```nix
defaultBindsFile = pkgs.runCommand "niri-default-binds.kdl" { } ''
  awk '/^binds/{f=1} f{print} f && /^}/{exit}' \
    ${cfg.package.src}/resources/default-config.kdl > $out
'';
```

and prepend `include "${defaultBindsFile}"` to the generated config, behind an opt-in option. Defaults then always match the user's niri version, per-key overrides via `settings.binds` just work, and there's no IFD since it's only a store-path reference. I've verified the awk extraction against the current upstream `default-config.kdl`, but not thoroughly vetted edge cases / possible gotchas.

File: tests/modules/services/niri/niri-empty-config.nix (hunk 2 above)
Comment (khaneliman) [reply tests/modules/services/niri/niri-empty-config.nix]: +1, and the `mkForce` needed here shows the current approach isn't backwards compatible: an existing `extraConfig = builtins.readFile ./config.kdl` setup would suddenly get the default binds prepended on top of its own `binds` section. Whatever shape this lands in, the defaults should be opt-in.

**Review by khaneliman — COMMENTED** [review]
Thanks for tackling this, silently ending up with zero keybindings is a real footgun.

My main concern is maintaining our own copy of upstream's default binds: we already have that problem in the sway module, where our defaults drift from upstream and surprise users. Here the drift also causes hard build failures (see inline). I've left a suggestion on the existing thread for sourcing version-matched defaults from the niri package itself instead.

File: modules/services/window-managers/niri.nix (hunk 1 above)
Comment (lukasngl) [reply modules/services/window-managers/niri.nix]: Good call, the include mechanism looks very clean, would we then even need to filter out the binds instead of including the whole defaults kdl? Also if upstream would rename the default config file, this would be a hard build error, should the default file be an option with default or YAGNI?

Comment (glmlm) [reply modules/services/window-managers/niri.nix]: > would we then even need to filter out the binds instead of including the whole defaults kdl?

I was actually thinking the exact same thing. The code has been updated, so I'd appreciate it if you could review again.

> Also if upstream would rename the default config file, this would be a hard build error, should the default file be an option with default or YAGNI?

This might be the only remaining open question.
My two cents: we should also think about the future maintenance burden and documentation bloat from adding an option that will rarely be used.

File: modules/services/window-managers/niri.nix [hunk 4]
```diff
@@ -143,6 +143,12 @@ in
       '';
     };
 
+    includeDefaultConfig = lib.mkOption {
```
Comment (lukasngl) [inline modules/services/window-managers/niri.nix]: I think the option should be enabled by default, otherwise the default is still the footgun.
Maybe consider using a enable option aswell? 
Something like:

```nix
enableDefaultConfig = (lib.mkEnableOption "including the default config from {option}`package`.") // { default = true; }
```

**Review by lukasngl — CHANGES_REQUESTED** [review]
Looks good! Some minor tweaks would be appreciated.

Have you tried it out and verified that it works end to end? I.e. there is no test case, i am pretty sure that the store path would be normalized to something like "0000...-niri/resources/default-config.kdl" in the test output, so maybe you could consider adding a test case? (I.e. flipping the default to true would require touching the testfiles anyways)

File: modules/services/window-managers/niri.nix (hunk 4 above)
Comment (glmlm) [reply modules/services/window-managers/niri.nix]: Thank you for your continued reviews!

Issue #9741 is about the lack of a way to retain the default config when using `extraConfig`, which may contradict user expectations. While I have no objection to changing the default value, this issue wouldn't occur if no config file exists since Niri automatically generates one on its first run.

> Maybe consider using a enable option aswell?

If our goal is simply to have it enabled by default, I wonder if we really need to introduce a new option. Could you share any use cases that can't be covered by just changing the default value of `includeDefaultConfig` to true?

**Comment by glmlm** [comment]
> Have you tried it out and verified that it works end to end?

Yes, since this is an essential feature for my own setup, I've been verifying each commit to ensure it actually works in real usage, rather than just passing the tests.

Regarding the test case you mentioned, hypothetically, would it still be preferable to update it even if we decide to keep the default value as false?

File: modules/services/window-managers/niri.nix (hunk 4 above)
Comment (lukasngl) [reply modules/services/window-managers/niri.nix]: Nah, you used a bare `lib.mkOption`, my suggestion was to use an `lib.mkEnableOption` and _renaming_ the option, as flags usually use this type and naming scheme.

**Comment by glmlm** [comment]
To facilitate our discussion, I have summarized the remaining open points below:

- The default value for `enableDefaultConfig`[[1]](https://github.com/nix-community/home-manager/pull/9745#discussion_r3695332532)[[2]](https://github.com/nix-community/home-manager/pull/9745#discussion_r3691365783)
- ~[Whether to add a new test case for this feature](https://github.com/nix-community/home-manager/pull/9745#pullrequestreview-4834391981)~
- [The necessity of introducing an option to customize the `default-config.kdl` source location](https://github.com/nix-community/home-manager/pull/9745#discussion_r3694775649)

File: modules/services/window-managers/niri.nix [hunk 5]
```diff
@@ -189,9 +191,11 @@ in
           escapeTabs = true;
         };
         settings = lib.removeSuffix "\n" (toKDL cfg.settings);
+        defaultConfig = lib.optionalString (cfg.enableDefaultConfig) ''include "${cfg.package.src}/resources/default-config.kdl"'';
```
Comment (khaneliman) [inline modules/services/window-managers/niri.nix]: issue (blocking): add coverage for the enabled path

Current tests leave `enableDefaultConfig` disabled, so green CI does not evaluate `cfg.package.src` or verify include ordering. Add a focused test with a stub `src`, assert the include precedes generated settings, and keep the empty-config behavior covered.

Comment (glmlm) [reply modules/services/window-managers/niri.nix]: While constructing the `package` beforehand feels a bit more idiomatic, I prioritized keeping the format consistent with the other tests.

For reference, the alternative approach I had in mind looks like this:

```nix
package =
  pkgs.runCommand "niri-stub-package"
    {
      src = pkgs.runCommand "niri-stub-source" { } ''
        mkdir -p $out/resources
        touch $out/resources/default-config.kdl
      '';
    }
    ''
      mkdir -p $out/bin
      touch $out/bin/niri
    '';
```

Since both patterns exist across other modules' tests, I'd appreciate your feedback on this.

Comment (khaneliman) [inline modules/services/window-managers/niri.nix]: issue (blocking): validate the nullable package before reading its source

`package` explicitly permits `null`, but enabling this option then evaluates `cfg.package.src`. Add this entry to `config.assertions` so Home Manager reports the invalid option combination clearly:

```nix
      {
        assertion = cfg.enableDefaultConfig -> cfg.package != null;
        message = "wayland.windowManager.niri.enableDefaultConfig requires a non-null package";
      }
```

File: modules/services/window-managers/niri.nix [hunk 6]
```diff
@@ -189,9 +195,11 @@ in
...
```
Comment (khaneliman) [inline modules/services/window-managers/niri.nix]: **[P2] Do not dereference `package.src` for arbitrary package values**

`package` uses `lib.mkPackageOption`, so raw store paths and package wrappers without a `src` attribute are valid values. When `enableDefaultConfig = true`, this interpolation evaluates `cfg.package.src` and fails with `attribute 'src' missing` before Home Manager generates the configuration. The new assertion rejects only `null`, so valid non-null custom packages still break evaluation.

File: modules/services/window-managers/niri.nix [hunk 7]
```diff
@@ -167,6 +169,12 @@ in
         assertion = cfg.checkConfig -> cfg.package != null;
         message = "wayland.windowManager.niri.checkConfig requires a non-null package";
       }
+      {
+        assertion =
+          cfg.enableDefaultConfig
+          -> (cfg.package != null && lib.hasAttr "src" cfg.package && cfg.package.src != null);
+        message = "wayland.windowManager.niri.enableDefaultConfig requires a non-null package with a valid `src` attribute";
```
Comment (khaneliman) [inline modules/services/window-managers/niri.nix]: blocking: make the assertion safe for every package value

`package` accepts context-bearing store-path strings. `lib.hasAttr` throws on them, so this assertion can crash instead of returning false. `? src` works on non-attribute values, and `lib.isStringLike` checks the value without realizing it or causing IFD.

```suggestion
        assertion =
          cfg.enableDefaultConfig
          -> (cfg.package ? src && lib.isStringLike cfg.package.src);
        message = "wayland.windowManager.niri.enableDefaultConfig requires a package with a string-like `src` attribute";
```

File: modules/services/window-managers/niri.nix [hunk 8]
```diff
@@ -189,9 +197,11 @@ in
           escapeTabs = true;
         };
         settings = lib.removeSuffix "\n" (toKDL cfg.settings);
+        defaultConfig = lib.optionalString (cfg.enableDefaultConfig) ''include "${cfg.package.src}/resources/default-config.kdl"'';
```
Comment (khaneliman) [inline modules/services/window-managers/niri.nix]: blocking: guard `.src` before `home.file` forces it

With `enableDefaultConfig = true` and `package = null`, the file assertions force this config while Home Manager is still collecting failed assertions. Evaluation dies at `.src` before it can print the assertion above. This guard keeps generation lazy, so the assertion can report the invalid combination.

```suggestion
        defaultConfig = lib.optionalString (
          cfg.enableDefaultConfig
          && cfg.package ? src
          && lib.isStringLike cfg.package.src
        ) ''include "${cfg.package.src}/resources/default-config.kdl"'';
```

Add a regression with `package = null` that expects the assertion message. That is the invalid combination this assertion was added to diagnose.

**Review by khaneliman — CHANGES_REQUESTED** [review]

File: modules/services/window-managers/niri.nix [hunk 9]
```diff
@@ -189,9 +195,13 @@ in
           escapeTabs = true;
         };
         settings = lib.removeSuffix "\n" (toKDL cfg.settings);
+        defaultConfig = lib.optionalString (
+          cfg.enableDefaultConfig && cfg.package ? src && lib.isStringLike cfg.package.src
```
Comment (khaneliman) [inline modules/services/window-managers/niri.nix]: issue: enableDefaultConfig emits include for niri versions before 25.11

`include` support starts in niri 25.11, but this condition emits the directive for any package with a string-like `src`, including niri 25.8. If a user sets `checkConfig = false`, Home Manager activates a config that the selected niri cannot parse. Gate this option on the niri version or provide a pre-25.11 fallback, and add a regression test.

Comment (khaneliman) [inline modules/services/window-managers/niri.nix]: issue (blocking): Reject packages that predate config includes

Niri added `include` in 25.11, but this condition emits the directive for any package with a string-like `src`, including niri 25.8. With `checkConfig = false`, Home Manager activates a configuration that the selected niri cannot parse. Either generate the selected package's default configuration without `include`, or assert that `enableDefaultConfig` requires a compatible niri version. Add a regression with a pre-25.11 package.
